### ISSUE TITLE
Generalize SERVICE for non-SPARQL endpoints

### ISSUE THREAD

**Issue by rubensworks** [w3c/sparql-12#10]
Currently, the `SERVICE` clause can only be used for querying SPARQL endpoints,
while the ability to query over non-SPARQL endpoints would also be useful.

## Why?

From the [SPARQL 1.1 federated query spec](https://www.w3.org/TR/sparql11-federated-query/#variableService):

> SERVICE calls depend on the SPARQL Protocol [SPROT] which transfers serialized RDF documents making blank nodes unique between service calls.

This allows queries like the following:

```sparql
SELECT *
WHERE {
  SERVICE <http://example.org/sparql> {
     ?s ?p ?o .
  } 
}
```

However, since RDF is being published on the Web as Linked Data
using various RDF serializations (Turtle, TriG, JSON-LD, ...),
it would be useful for querying these with the `SERVICE` keyword as well.

This would enable queries like the following:

```sparql
SELECT *
WHERE {
  SERVICE <http://example.org/me.rdf> {
     ?s ?p ?o .
  } 
}
```

This would require a SPARQL engine to:

1. Fetch the remote document (_with conneg for supported RDF serializations_)
3. Parse the document based on the document's content type to in-memory RDF.
4. Query over the in-memory RDF.

## Previous work

The [Comunica framework](http://comunica.linkeddatafragments.org/) automatically detects the type of source in `SERVICE` clauses,
and queries the source appropriately.

## Considerations for backward compatibility

This would be fully backwards-compatible, as existing `SERVICE` clauses for SPARQL endpoints would still work without changes to the query string.

**Comment by VladimirAlexiev** [comment]
Isn't http://example.org/sparql without `?query` supposed to return the SPARQL Service Description RDF? Methinks:
- you need a new keyword for "go slurp that file"
- you need to be able to specify the Accept header, in case the file is served with content negotiation

**Comment by frmichel** [comment]
+1.
Related work can also be found in SPARQL-LD [1].

[1] Fafalios P., Yannakis T. & Tzitzikas Y. (2016). Querying the Web of Data with SPARQL-LD. In Research and Advanced Technology for Digital Libraries vol. 9819, Fuhr, N., Kovács, L., Risse, T. & Nejdl, W. (Eds.), pp. 175–187. Cham. Springer International Publishing.

**Comment by VladimirAlexiev** [comment]
The paper is available from https://www.semanticscholar.org/paper/Querying-the-Web-of-Data-with-SPARQL-LD-Fafalios-Yannakis/6dc4883228c95e8a332320fcc587a0ff33c84d59. Further relevant citations:
- Heuristics-based Query Reordering for Federated Queries in SPARQL 1.1 and SPARQL-LD
Thanos Yannakis, Pavlos Fafalios, Yannis Tzitzikas. GeoLD-QuWeDa@ESWC2018
- How Many and What Types of SPARQL Queries can be Answered through Zero-Knowledge Link Traversal? Pavlos Fafalios, Yannis Tzitzikas. ArXiv2018

**Comment by seralf** [comment]
Hi I think we may consider also some of the existing approaches when using full-text search.

for example in the blazegraph documentation:
https://wiki.blazegraph.com/wiki/index.php/FullTextSearch

we can find either examples with the adoption of a custom usage of the `SERVICE` keyword to enable custom (full-text) functions, such as:
```sparql
PREFIX bds: <http://www.bigdata.com/rdf/search#>
SELECT ?sub ?score
 WHERE {
   SERVICE <http://www.bigdata.com/rdf/search#search> {
     ?lit bds:search "mike~2" .
     ?lit bds:relevance ?score .
   }
   ?subj ?p ?lit .
}
```
or on the other hand: https://wiki.blazegraph.com/wiki/index.php/ExternalFullTextSearch
the adoption of specific XXX keyword to enable calling external (elasticsearch, solr, and so on) endpoints, such as:
```sparql
PREFIX fts: <http://www.bigdata.com/rdf/fts#>
SELECT ?res ?score ?snippet WHERE {
  ?res fts:search "Alice | Bob" .
  ?res fts:endpoint "http://localhost:1234/solr/blazegraph/select" .
  ?res fts:endpointType  "SOLR" .
  ?res fts:score ?score .
  ?res fts:fieldToSearch "text" .
  ?res fts:scoreField "score" .
}
```
I think a syntax that is a mix of the two previous examples could be a good solution for those cases.

Handling all the elements needed for search inside the block defined for a `SERVICE` could also have some benefit in terms of possible correct analysis and plan for the query, I suppose.

**Comment by osma** [comment]
The Wikidata SPARQL endpoint ([WDQS](https://query.wikidata.org/)) also has a [Label service](https://www.mediawiki.org/wiki/Wikidata_Query_Service/User_Manual#Label_service) that works via SERVICE even though it isn't actually a SPARQL endpoint.

Example (from above linked page):
```
SELECT ?p ?pLabel ?w ?wLabel WHERE {
   wd:Q30 p:P6/ps:P6 ?p .
   ?p wdt:P26 ?w .
   SERVICE wikibase:label {
    bd:serviceParam wikibase:language "en" .
   }
 }
```

The service understands that e.g. `?pLabel` should be bound to a suitable label of `?p` (and similar for `?w`/`?wLabel`) even though it isn't stated in the query.

A bit ugly maybe, but this fulfills a real need. The standard mechanism for asking for labels via triple patterns and `FILTER` on language tags is too inefficient for Wikidata, which is extremely multilingual, especially if you need a fallback mechanism in case a label in the preferred language doesn't exist.

**Comment by cygri** [comment]
@osma I wonder why Wikidata's label service wasn't done as an extension function:
```
SELECT ?p ?pLabel ?w (wikibase:label(?w, "en") AS ?wLabel) {
   wd:Q30 p:P6/ps:P6 ?p .
   ?p wdt:P26 ?w .
   BIND (wikibase:label(?p, "en") AS ?pLabel)
}
```
The example shows two different ways of using the hypothetical extension function—in the `SELECT` clause and with `BIND`. This would use an existing extension point of the language and the result is more idiomatic and flexible. I assume there is a downside that I'm not aware of.

**Comment by afs** [comment]
This is an high level outline to test how to proceed.

There is a connection to issue #6 (Keyword for functions that can produce multiple bindings) and others.

There could be a single mechanism for extensions named by URI that takes arguments and returns a result table. 

There are important details to work on such as whether and how the callout declares its "in"and "out" parameters and whether the "in" parameters must be.

There may be specific syntax for important cases to make the mechanism easier to use for those cases and aid engines providing only certain common cases.

#6 - Keyword for functions that can produce multiple bindings
#10 - Generalize SERVICE for non-SPARQL endpoints
#40 - Standardize free text search of RDF data
#44 - Named solution sets
[R2RML](https://www.w3.org/TR/r2rml/)
[CSVW](https://www.w3.org/TR/tabular-data-primer/)

**Comment by osma** [comment]
@cygri I don't know why Wikidata chose SERVICE for the label service. I agree that an extension function would seem appropriate.

Maybe @smalyshev could shed some light on this?

**Comment by seralf** [comment]
> There could be a single mechanism for extensions named by URI that takes arguments and returns a result table.
> There are important details to work on such as whether and how the callout declares its "in"and "out" parameters and whether the "in" parameters must be.

I wonder if for this could be taken as a reference the work with [function ontology](https://fno.io/)?
This approach is already used in the context of R2RML proposed extensions (for example on: https://github.com/carml/carml#function-extension and https://www.slideshare.net/christophedebruyne/r2rmlf-towards-sharing-and-executing-domain-logic-in-r2rml-mappings)
and seems to be a good starting point to analyze requirements and pros/cons of a possible syntax

**Comment by kasei** [comment]
I would love to see us develop new features that can help avoid the use of `SERVICE` for non-standard, and implementation-specific extensions such as done in wikidata. There's already discussion in #13 about how to address the specific case of `wikibase:label` and `wikibase:language`.

**Comment by smalyshev** [comment]
@osma SERVICE is the only extension syntax in SPARQL that can be used for this AFAIK. Function does not naturally accept multiple output parameters, while SERVICE clause can have multiple outputs. Also, it'd be harder to do rewrites that label service does with functions, IMHO - such as ones described here: https://www.mediawiki.org/wiki/Wikidata_Query_Service/User_Manual#Label_service where SERVICE clause is used for instantiating the labels, we specify language only once and it works for all labels. Maybe it'd be possible to use functions with different syntax of course. 

@kasei It would definitely be nice to have some extension point that can be used as SERVICE is used now - for providing extended functionality that is not always covered by functions - e.g. many optional parameters, multiple outputs, etc. Wikidata has more extensions than labels (see https://www.mediawiki.org/wiki/Wikidata_Query_Service/User_Manual#Extensions) - I plan to add the list to https://github.com/w3c/sparql-12/wiki/Inventory-of-existing-extensions-to-SPARQL-1.1 shortly.

**Comment by osma** [comment]
@smalyshev Thanks for the explanation!

[jena-text](https://jena.apache.org/documentation/query/text-query.html#query-with-sparql) uses a property function `text:query` which *does* have multiple outputs. The most general form is this:

```
 ( ?s ?score ?literal ?g ) text:query ( property 'query string' limit 'lang:xx' 'highlight:yy' )
```

although you can also use a simpler form with a single output, like this:

```
?s text:query 'word' 
```

Not saying this is better, just that it's possible, at least in Jena ARQ, to have multiple outputs for a  property function.

**Comment by cygri** [comment]
@smalyshev @osma We have #6 for working towards a dedicated syntax for multi-result functions.

**Comment by kasei** [comment]
@smalyshev 

> SERVICE is the only extension syntax in SPARQL that can be used for this AFAIK

I'd quibble a bit with the idea that `SERVICE` is an "extension syntax." However, along these same lines, it seems to me that a 'magic' named `GRAPH` would have satisfied the same requirements for wikidata?

However, my biggest objection to the way in which wikidata uses `SERVICE` is in specific extensions such as `wikibase:label` which introduce non-local effects on the entire query, and cause variables to be bound which do not appear anywhere in the graph pattern. That's seems to me like a code smell that should make us consider the problem that's being solved, and find a way to address it directly.

**Comment by smalyshev** [comment]
> specific extensions such as wikibase:label which introduce non-local effects on the entire query, and cause variables to be bound which do not appear anywhere in the graph pattern.

This is just syntax sugar, this query is rewritten to include these variables in the actual pattern the engine sees, it's just not visible to the user. I am not sure where "syntax sugar" features fall on the spectrum of features. 

> That's seems to me like a code smell that should make us consider the problem that's being solved, and find a way to address it directly.

The problem is to allow users to address a very common use case directly, namely:
* The query returns a set of entities which have non-human-friendly URIs
* These entities have multi-lingual human-friendly labels, which the user wants to see in the result
* The user usually wants the labels in one specific language, but they would also want fallbacks - i.e.  if one speaks French, French label is the best, but if you don't have a French label, English one may do, if you don't even have English label, display _something_ at least that allows us to know which entity it is (i.e. not blank and not dropping the row). 
* There may be many such entities in the query result, so the amount of boilerplate code per entity should be minimized
* The language preference is usually common between all the entities, so it should be only specified once
* There are also descriptions, which may be thought of for the purpose of this use case as "longer, more descriptive labels", that are needed less frequently, but when they do, they require the same treatment
* The syntax for doing this should be easy to remember and reproduce for a beginner (maybe with some copy-pasting) since most beginner queries would need this

**Comment by kasei** [comment]
@smalyshev 

> This is just syntax sugar, this query is rewritten to include these variables in the actual pattern the engine sees, it's just not visible to the user. I am not sure where "syntax sugar" features fall on the spectrum of features.

As we've discussed before, I think this is a really valuable use case to be solving. My concern is that while the wikidata solution may be good for newcomers *to SPARQL*, it is probably not as good for newcomers *to wikidata* (who already have some understanding of SPARQL), and it makes interop with other SPARQL implementations much harder (it's not a normal extension point in many implementations, and requires global rewriting contrary to the syntax alone which would just imply runtime operations).

As an aside, I had a hard time figuring out what sort of syntactic rewriting is going on with `wikibase:label`. I tried several queries that used the automatically bound variables in ways I would expect to work, but failed to make sense of it. Is there a way to see the result of this rewriting process? (Being able to access the rewritten query might also benefit interop attempts.)

**Comment by cygri** [comment]
@smalyshev I'm wondering if the approach I'm proposing at #6 would work for this use case? If a user wanted to get labels for the items in three variables `?p` `?w` and `?o`, and also the descriptions for `?o`, all in French with English as fallback, they could add this to the end of the query:
```
VALUES * OF wikibase:labels("fr,en", ?p, ?w, ?o)
VALUES * OF wikibase:descriptions("fr,en", ?o)
```
The implementation of `wikibase:labels` and `wikibase:descriptions` would fetch the respective values and bind them to `?pLabel`, `?wLabel`, `?oLabel`, and `?oDescription`. If the user wanted other variable names, they could write:
```
VALUES (?pname ?wname ?oname) OF wikibase:labels("fr,en", ?p, ?w, ?o)
```

**Comment by enridaga** [comment]
Generalising the definition of the SERVICE clause in the specification may be sufficient, leaving providers to support different ways of injecting complex behaviour in the SPARQL query.

An advanced version is to include in the specification methods for customising the SERVICE request, as pointed out in https://github.com/w3c/sparql-12/issues/10#issuecomment-480199950

In [SPARQL Anything](https://github.com/SPARQL-Anything/sparql.anything) we "override" the SERVICE clause with a new protocol `x-sparql-anything:` and define a vocabulary for passing configurations, similarly to the technique typically used for full-text searches. One example: https://github.com/SPARQL-Anything/sparql.anything/issues/65#issuecomment-886909666

**Comment by TallTed** [comment]
Note that "querying" remote static files as described above can be horribly inefficient if those files are large, especially if there are multiple references to them in your query -- which may cause a large `.ttl` or other serialized RDF file to be transferred over the wire multiple times!

Virtuoso delivers the apparently desired results (which do not necessarily require misuse of `SERVICE` like what Wikidata has done, so I would suggest retitling this issue and rewriting the OP summary to focus more on the desired net-result and focus less on how that is achieved) with a combination of extension of `FROM NAMED` with `OPTION` and Sponger ingestion, as [seen here](http://docs.openlinksw.com/virtuoso/rdfinsertmethodsparqlqueryandsponger/) --

```
SELECT ?id
FROM NAMED <http://www.openlinksw.com/dataspace/kidehen@openlinksw.com/weblog/kidehen@openlinksw.com%27s%20BLOG%20%5B127%5D/sioc.ttl>
 OPTION (get:soft "soft", get:method "GET")
WHERE { GRAPH ?g { ?id a ?o } }
LIMIT 10
```

There's [more here](http://docs.openlinksw.com/virtuoso/rdfinputgrab/).

**Comment by VladimirAlexiev** [comment]
This issue is a big mish-mash of valuable and important ideas.

I'll add another:
- IMHO the most powerful way to dynamically bring in external data (RDF or not) and transform it is [SPARQL-Generate](https://ci.mines-stetienne.fr/sparql-generate/).
- (It's also used to extract and generate non-RDF out of RDF data)

**Comment by namedgraph** [comment]
SPARQL-Generate and tools like that are basically a language on its own, with a single implementation, because no standard SPARQL tools support it. So I'm not sure how useful it is to them as an example.

**Comment by VladimirAlexiev** [comment]
@namedgraph The relevance to this issue is that people started discussing the various uses that SERVICE has been put in, examples including FTS, BlazeGraph extensions, etc.

Here's another: [GraphDB Path Search](https://graphdb.ontotext.com/documentation/standard/graph-path-search.html) uses SERVICE:
- (outer) To invoke the search algorithm (like Blazegraph and others do)
- (inner) To specify the triple pattern for each step (where you can use Sparql Property Paths or any other patterns)

Eg here's a search of actors related through films (your typical [Six Degrees of Kevin Bacon](https://en.wikipedia.org/wiki/Six_Degrees_of_Kevin_Bacon) problem):
```sparql
PREFIX path: <http://www.ontotext.com/path#>
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?edge ?index ?path
WHERE {
    VALUES (?src ?dst) {
        ( <http://dbpedia.org/resource/Chris_Evans_(actor)> <http://dbpedia.org/resource/Chris_Hemsworth> )
    }
    SERVICE <http://www.ontotext.com/path#search> {
        <urn:path> path:findPath path:allPaths ;
                   path:sourceNode ?src ;
                   path:destinationNode ?dst ;
                   path:pathIndex ?path ;
                   path:minPathLength 2 ;
                   path:startNode ?start;
                   path:resultBinding ?edge ;
                   path:endNode ?end;
                   path:resultBindingIndex ?index .
        SERVICE <urn:path> {
            ?film a <http://dbpedia.org/ontology/Film> .
            ?film <http://dbpedia.org/property/starring> ?start .
            ?film <http://dbpedia.org/property/starring> ?end .
        }
    }
}
```

A lot of this better belongs to https://github.com/w3c/sparql-12/wiki/Inventory-of-existing-extensions-to-SPARQL-1.1

**Comment by afs** [comment]
Personal request - please be cautious when broadening an issue. If it is a collection of things, early contributions before the divergence, tend to get lost. Please do open issues when diverging and use links.

**Comment by afs** [comment]
For extension functionality, local algorithms etc, we are not restricted by current (1.1) syntax : see issue #6.

* Declared arguments as to whether in-required, in-optional, out-always and out-optional.
* Named argument syntax ideas for readable many-argument cases.
* Iterate vs join (virtual table) semantics.

I believe that the addition of syntax to call custom processing is one of the important items to do because it has general applicability to several areas. It is best if both client software and server software understand the syntax, even if they do not support the specific extension.

**Comment by Aklakan** [comment]
If it was standardize to allow `SERVICE <http://example.org/me.rdf>` then we'd need standardized HTTP CONNEG options as well.

Example:

>  "The Comunica framework automatically detects the type of source in SERVICE clauses,
and queries the source appropriately."

How would communica know that someone wanted the `me.rdf.ttl` representation in preference to the outdated `me.rdf.xml` one ;)

**Comment by dbooth-boston** [comment]
_BRIEF META COMMENT:  I just read the above discussion, and it is impressive how constructive, knowledgeable and respectful the comments are.  This is an excellent example of the kind of community that makes people want to contribute.  Nice job folks!_

**Comment by VladimirAlexiev** [comment]
@dbooth-boston I second this: this WG (github issues) is a burgeoning forum of excellent ideas. Even if not all are directly about SPARQL, they are very relevant and stimulating.
